plugins/producer_plugin/producer_plugin.cpp

```diff
@@ -2829,98 +2856,126 @@ void producer_plugin_impl::start_write_window() {
+         exhausted_trx_queue_empty = !_ro_exhausted_trx_queue.pop_front(t);
+         if (!exhausted_trx_queue_empty) {
+            auto retry = process_read_only_transaction( t.trx, t.next );
+            if( retry ) {
+               _ro_exiting_read_window = true;
+               break;
+            }
+         }
+      } else {
+         bool more = app().executor().execute_highest_read_only();
+         if ( !more ) { // read_only queue empty, anything queued after this will be processed in the next window
+            _ro_exiting_read_window = true;
+            break;
+         }
       }
    }
 
    // If all tasks are finished, do not wait until end of read window; switch to write window now.
-   if ( --_ro_num_active_trx_exec_tasks == 0 ) {
-      // Do switching on app thread to serialize
-      app().executor().post( priority::high, exec_queue::read_only_trx_safe, [self=this]() {
-         self->_ro_trx_exec_tasks_fut.clear();
+   if ( --_ro_num_active_exec_tasks == 0 ) {
+      // Needs to be on read_only because that is what is being processed until switch_to_write_window().
+      app().executor().post( priority::high, exec_queue::read_only, [self=this]() {
+         self->_ro_exec_tasks_fut.clear();
+         // will be executed from the main app thread because all read-only threads are idle now
          self->switch_to_write_window();
       } );
    }
 
    return true;
 }
 
+// Called from app thread during start block.
+// Used for executing exhausted read-only trxs when no read only thread pool enabled.
+void producer_plugin_impl::process_read_only_transactions(const fc::time_point& deadline) {
+   // if the read only thread pool is enabled then exhausted read only transactions will be executed
+   // by read_only_execution_task, so don't execute them here.
+   if ( _ro_thread_pool_size == 0 ) {
+      _ro_window_deadline = std::min(deadline, fc::time_point::now() + _ro_read_window_effective_time_us);
+      bool exhausted_trx_queue_empty = false;
+      while ( !exhausted_trx_queue_empty && fc::time_point::now() < _ro_window_deadline && !_received_block ) {
+         ro_trx_t t;
+         exhausted_trx_queue_empty = !_ro_exhausted_trx_queue.pop_front(t);
+         if (!exhausted_trx_queue_empty) {
+            auto retry = process_read_only_transaction( t.trx, t.next );
+            if( retry ) {
+               _ro_exiting_read_window = true;
+               break;
+            }
+         }
+      }
+   }
+}
+
 // Called from a read only trx thread. Run in parallel with app and other read only trx threads
 // Return whether the trx needs to be retried in next read window
-bool producer_plugin_impl::process_read_only_transaction(const packed_transaction_ptr& trx, const next_function<transaction_trace_ptr>& next, const fc::time_point& read_window_start_time) {
+bool producer_plugin_impl::process_read_only_transaction(const packed_transaction_ptr& trx, const next_function<transaction_trace_ptr>& next) {
```

Comment:
`next` was changed to a `const&` because it is normally not moved/copied, only used.

Reply:
It could take `ro_trx` by value. Would require creating a `ro_trx` in `on_incoming_transaction_async` but we are creating affectively the same thing. I would lean toward not doing that because I feel `ro_trx_t` is more of an implementation type and hence would prefer to keep it out of function signatures. But I don't feel strongly either way. I'll switch to by value if you want?